src/main/gui/main-window.ts

```diff
@@ -126,39 +125,39 @@ const registerEventHandlerPreSetup = (
         mainWindow.webContents.send('window-blur');
     });
 
-    if (isMac) {
-        // Open file with chaiNNer on macOS
-        app.on('open-file', (event, filePath) => {
-            (async () => {
-                const file = filePath;
-                if (file) {
-                    const result = await openSaveFile(file);
-                    mainWindow.webContents.send('file-open', result);
-                }
-            })().catch(log.error);
-        });
-    } else {
-        // Opening file with chaiNNer on other platforms
-        if (args.file) {
-            const result = openSaveFile(args.file);
-            ipcMain.handle('get-cli-open', () => result);
-        } else {
-            ipcMain.handle('get-cli-open', () => undefined);
-        }
+    // Open files with chaiNNer on macOS while the app is running
+    app.on('open-file', (event, filePath) => {
+        event.preventDefault();
+        (async () => {
+            const result = await openSaveFile(filePath);
+            mainWindow.webContents.send('file-open', result);
+        })().catch(log.error);
+    });
 
-        app.on('second-instance', (_event, commandLine) => {
-            (async () => {
-                const { file } = parseArgs(commandLine.slice(app.isPackaged ? 2 : 3));
-                if (file) {
-                    const result = await openSaveFile(file);
-                    mainWindow.webContents.send('file-open', result);
-                }
-                // Focus main window if a second instance was attempted
-                if (mainWindow.isMinimized()) mainWindow.restore();
-                mainWindow.focus();
-            })().catch(log.error);
-        });
+    if (args.file) {
+        // Open file with chaiNNer on other platforms
+        const result = openSaveFile(args.file);
+        ipcMain.handle('get-cli-open', () => result);
+    } else if (process.env.STARTUPFILE) {
+        // Open file with chaiNNer on macOS when app was closed
+        const result = openSaveFile(process.env.STARTUPFILE);
+        ipcMain.handle('get-cli-open', () => result);
```

Comment:
This is a race condition. You are racing the app loading enough to register events and the `app.on('open-file'` in `src/main/main.ts`.

Reply:
How do you mean? The `app.on` event in main gets always executed first and just sets the environment variable.

I tried different approaches that is the only one that works.

Reply:
> The `app.on` event in main gets always executed first

There is no guarantee that this happens. `app.on` is for asynchronous events, so `open-file` may be fired whenever. Of course, it will likely be fired near the beginning of the application start up, but that doesn't solve anything. When the main window loads quickly enough, when we speed up the start-up process in the future, or when we register other events sooner, this will stop working.

You are relying on the order of 2 asynchronous events without any synchronization mechanism between them. That's a race condition.

Reply:
Thanks for the feedback. But how should this be done, then? The documentation says on macOS, one needs to listen for this event before the `app.on('ready'`.
[https://www.electronjs.org/docs/latest/api/app#event-open-file-macos](https://www.electronjs.org/docs/latest/api/app#event-open-file-macos)

There are only two spots where I could do this. The first is in the `main.ts` (as per this PR) and the second one is `create.ts`. If I listen to this event later (like we do currently), then the event has already disappeared.